Refresh custom filter menu (#2595)
telerik/blazor-docs: knowledge-base/grid-custom-filter-menu.md

---

> title: How to Refresh Filter Menu Context After Programmatic Changes

Reviewer: Aren't we refreshing the Filter Menu UI? The context is always up-to-date, but the users don't see it.

description: Learn how to refresh filter menu context after programmatic changes with a custom component
type: how-to
page_title: How to Refresh Filter Menu Context After Programmatic Changes
slug: grid-kb-custom-filter-menu

> tags: grid, blazor, custom, filter, filtermenu

Reviewer: Tags should make sense outside of context. See the forums tags in the right column here: https://www.telerik.com/forums/blazor and imagine you see a "custom" tag there.

```suggestion
tags: blazor, grid, filter, filtermenu
```

res_type: kb
ticketid: 1669381
---

## Environment
<table>
    <tbody>
	    <tr>
	    	<td>Product</td>
	    	<td>Grid for Blazor</td>
	    </tr>
    </tbody>
</table>

## Description

> When using the TelerikGrid in Blazor applications, it's common to implement custom `FilterMenu` for enhanced filtering capabilities. However, integrating an [AutoComplete]({%slug autocomplete-overview%}) component with `StringOperators` within a [`FilterMenuTemplate`]({%slug grid-templates-filter%}#filter-menu-template) for dynamic filtering based on user input does not work as expected. The AutoComplete component does not consider the selection from the dropdown list of `StringOperators` within the `FilterMenuTemplate`. 

Reviewer: This paragraph makes it sound like only the AutoComplete poses a problem. You can leave the specific information in the KB, but consider starting with a more general description of the problem.

This knowledge base article also answers the following questions:
- How to integrate AutoComplete with `StringOperators` in a TelerikGrid `FilterMenuTemplate`?

> - How to refresh a custom `FilterMenuTemplate` in TelerikGrid?

Reviewer: The content of every template is "custom", so no need to explicitly mention this.

```suggestion
- How to refresh the contents of `FilterMenuTemplate` in a TelerikGrid for Blazor?
```

> - How to refresh a `FilterMenuTemplate` context after programmatic changes?

Reviewer: Same note as for the KB title.

## Solution

> To achieve the desired behavior, encapsulate the content of the `FilterMenuTemplate` in a custom component and refresh this component upon selection change in the dropdown list. Below is an example demonstrating this approach:

Reviewer: ```suggestion
To achieve the desired behavior, encapsulate the content of the `FilterMenuTemplate` in a separate Razor component and refresh this component upon selection change in the dropdown list. Below is an example demonstrating this approach:
```

<div class="skip-repl"></div>
````Home.razor

> @using Telerik.DataSource

Reviewer: ```suggestion
@using Telerik.DataSource

```

             FilterMode="GridFilterMode.FilterMenu"
             PageSize="25">
    <GridColumns>
        <GridColumn Field="@(nameof(Country.CountryName))">
            <FilterMenuTemplate>
                <CustomFilterMenu @bind-FilterOperator="@FilterOperator"
                                  @bind-SelectedCountry="@SelectedCountry"
                                  Countries="@Countries"
                                  FilterOperators="@FilterOperators" />
            </FilterMenuTemplate>
            <FilterMenuButtonsTemplate Context="filterContext">
                <TelerikButton OnClick="@(async _ => await FilterAsync(filterContext))" Class="k-button-solid-primary">Filter</TelerikButton>
                <TelerikButton OnClick="@(async _ => await ClearFilterAsync(filterContext))">Clear</TelerikButton>
            </FilterMenuButtonsTemplate>
        </GridColumn>
    </GridColumns>
</TelerikGrid>

@code {
    private IEnumerable<StringFilterOperator> FilterOperators { get; set; } = Enum.GetValues(typeof(StringFilterOperator)).Cast<StringFilterOperator>().ToList();
    private StringFilterOperator FilterOperator { get; set; } = StringFilterOperator.StartsWith;

>     private string SelectedCountry { get; set; } = null!;

Reviewer: `null!` is cheating in this case, because the value is not populated all the time.

```suggestion
    private string SelectedCountry { get; set; } = string.Empty;
```

    private async Task FilterAsync(FilterMenuTemplateContext filterContext)
    {
        FilterHandler(filterContext.FilterDescriptor);
        await filterContext.FilterAsync();
    }

    private async Task ClearFilterAsync(FilterMenuTemplateContext filterContext)
    {
        SelectedCountry = string.Empty;

        FilterHandler(filterContext.FilterDescriptor);
        filterContext.FilterDescriptor.FilterDescriptors.Clear();
        await filterContext.ClearFilterAsync();
    }

>     private void FilterHandler(CompositeFilterDescriptor filterDescriptor)

Reviewer: Consider another more descriptive name for this method. For example, `AddFilterDescriptor` or whatever you prefer.

    {
        var model = string.Empty;
        object? value = null;

        value = SelectedCountry;
        model = nameof(Country.CountryName);

        filterDescriptor.FilterDescriptors.Clear();

>         filterDescriptor.FilterDescriptors.Add(new FilterDescriptor(model, Telerik.DataSource.FilterOperator.IsEqualTo, value));

Reviewer: The operator should not be hard-coded here.

    }

    private List<Country> Countries { get; set; } = new List<Country>()
{

>     new Country { Id = 1, CountryName = "Albania" },

Reviewer: Consider random data that is generated in a loop. Will make the code a lot shorter. In this case, it makes sense to have a different string in each data item and this is possible too. If you ensure that every letter and digit is represented, you will be fine. For example - https://docs.telerik.com/blazor-ui/components/grid/filter/searchbox

Current filter operator: @FilterOperator

<TelerikDropDownList Value="@FilterOperator"
                     ValueChanged="@( (StringFilterOperator newValue) => OnFilterOperatorChanged(newValue) )"
                     Data="@FilterOperators"
                     Width="160px">
</TelerikDropDownList>

<TelerikAutoComplete ScrollMode="@DropDownScrollMode.Virtual"
                     Value="@SelectedCountry"
                     ValueChanged="@( (string selectedCountry) => HandleSelectedCountryChange(selectedCountry) )"
                     TItem="Country"
                     Data="@Countries"
                     ValueField="CountryName"
                     Width="300px"
                     PageSize="20"
                     Filterable="true"
                     ItemHeight="35"
                     FilterOperator="@FilterOperator">
</TelerikAutoComplete>

@code {
    [Parameter] public EventCallback<StringFilterOperator> FilterOperatorChanged { get; set; }
    [Parameter] public EventCallback<string> SelectedCountryChanged { get; set; }
    [Parameter] public StringFilterOperator FilterOperator { get; set; }
    [Parameter] public string SelectedCountry { get; set; } = null!;

>     [Parameter] public IEnumerable<Country> Countries { get; set; } = null!;

Reviewer: ```suggestion
    [Parameter] public IEnumerable<Country> Countries { get; set; } = Enumerable.Empty<Country>();
```

    new Country { Id = 34, CountryName = "Netherlands" },
    new Country { Id = 35, CountryName = "Norway" },
    new Country { Id = 36, CountryName = "Poland" },
    new Country { Id = 37, CountryName = "Portugal" },
    new Country { Id = 38, CountryName = "Romania" },
    new Country { Id = 39, CountryName = "Russia" },
    new Country { Id = 40, CountryName = "San Marino" },
    new Country { Id = 41, CountryName = "Serbia" },
    new Country { Id = 42, CountryName = "Slovakia" },
    new Country { Id = 43, CountryName = "Slovenia" },
    new Country { Id = 44, CountryName = "Spain" },
    new Country { Id = 45, CountryName = "Sweden" },
    new Country { Id = 46, CountryName = "Switzerland" },
    new Country { Id = 47, CountryName = "Turkey" },
    new Country { Id = 48, CountryName = "Ukraine" },
    new Country { Id = 49, CountryName = "United Kingdom" },
    new Country { Id = 50, CountryName = "Vatican City" }
};
    public class Country
    {
        public int Id { get; set; }
        public string CountryName { get; set; } = null!;
    }
}
````
````CustomFilterMenu.razor
@using Telerik.DataSource
@using Microsoft.AspNetCore.Components
@using static Home

> <label for="filterOperator"><b>Select Filter Operator for the AutoComplete:</b></label>

Reviewer: Use `<strong>` instead of `<b>`. The former is semantic markup, which implies importance, while the latter is just a style.
